xrdp/xrdp_mm.c

```diff
@@ -991,11 +991,185 @@ xrdp_mm_process_rail_drawing_orders(struct xrdp_mm *self, struct stream *s)
+            in_uint32_le(s, Left);
+            in_uint32_le(s, Top);
+            in_uint32_le(s, Width);
+            in_uint32_le(s, Height);
+            in_uint32_le(s, PhysicalWidth);
+            in_uint32_le(s, PhysicalHeight);
+            in_uint32_le(s, Orientation);
+            in_uint32_le(s, DesktopScaleFactor);
+            in_uint32_le(s, DeviceScaleFactor);
+            LOG(LOG_LEVEL_INFO, "    Flags 0x%8.8x Left %d Top %d "
+                   "Width %d Height %d PhysicalWidth %d PhysicalHeight %d "
+                   "Orientation %d DesktopScaleFactor %d DeviceScaleFactor %d",
+                   Flags, Left, Top, Width, Height,
+                   PhysicalWidth, PhysicalHeight, Orientation,
+                   DesktopScaleFactor, DeviceScaleFactor);
+
+            rect.left = MIN(Left, rect.left);
+            rect.top = MIN(Top, rect.top);
+            rect.right = MAX(rect.right, Left + Width);
+            rect.bottom = MAX(rect.bottom, Top + Height);
+        }
+    }
+    session_width = rect.right - rect.left;
+    session_height = rect.bottom - rect.top;
+    if ((session_width > 0) && (session_height > 0))
+    {
+        // TODO: Unify this logic with server_reset
+        libxrdp_reset(wm->session, session_width, session_height, 32);
+        /* reset cache */
+        xrdp_cache_reset(wm->cache, wm->client_info);
+        /* resize the main window */
+        xrdp_bitmap_resize(wm->screen, wm->client_info->width,
+                           wm->client_info->height);
+        /* load some stuff */
+        xrdp_wm_load_static_colors_plus(wm, 0);
+        xrdp_wm_load_static_pointers(wm);
+        /* redraw */
+        xrdp_bitmap_invalidate(wm->screen, 0);
+
+        struct xrdp_mod* v = wm->mm->mod;
+        if (v != 0) {
+            v->mod_server_version_message(v);
+            v->mod_server_monitor_resize(v, session_width, session_height, wm->screen->bpp);
+            v->mod_server_monitor_full_invalidate(v, session_width, session_height);
+        }
+    }
+    return 0;
+}
+
 /******************************************************************************/
 int
 xrdp_mm_drdynvc_up(struct xrdp_mm *self)
 {
     LOG_DEVEL(LOG_LEVEL_INFO, "xrdp_mm_drdynvc_up:");
+    struct xrdp_drdynvc_procs d_procs;
+    int flags;
+    int error;
+
+    g_memset(&d_procs, 0, sizeof(d_procs));
+    d_procs.open_response = dynamic_monitor_open_response;
+    d_procs.close_response = dynamic_monitor_close_response;
+    d_procs.data_first = dynamic_monitor_data_first;
+    d_procs.data = dynamic_monitor_data;
+    flags = 0;
+    error = libxrdp_drdynvc_open(self->wm->session,
+                                 "Microsoft::Windows::RDS::DisplayControl",
+                                 flags, &d_procs,
+                                 &(self->dynamic_monitor_chanid));
+    if (error != 0)
+    {
+        LOG_DEVEL(LOG_LEVEL_INFO, "xrdp_mm_drdynvc_up: libxrdp_drdynvc_open failed %d", error);
```

Comment:
Definitely needs a LOG() as this is something the user needs to report. Suggest LOG_LEVEL_ERROR.